manifests/renew.pp

```diff
@@ -0,0 +1,98 @@
+# == Class: letsencrypt::renew
+#
+#   This class configures renewal of Let's Encrypt certificates using the
+#   certbot renew command.
+#
+#   Note: Hooks set here will run before/after/for ALL certificates, including
+#   any not managed by Puppet. If you want to create hooks for specific
+#   certificates only, create them using letsencrypt::certonly.
+#
+# === Parameters:
+#
+# [*pre_hook_commands*]
+#   Array of commands to run in a shell before obtaining/renewing any certificates.
+# [*post_hook_commands*]
+#   Array of commands to run in a shell after attempting to obtain/renew certificates.
+# [*deploy_hook_commands*]
+#   Array of commands to run in a shell once for each successfully issued/renewed
+#   certificate. Two environmental variables are supplied by certbot:
+#   - $RENEWED_LINEAGE: Points to the live directory with the cert files and key.
+#                       Example: /etc/letsencrypt/live/example.com
+#   - $RENEWED_DOMAINS: A space-delimited list of renewed certificate domains.
+#                       Example: "example.com www.example.com"
+# [*additional_args*]
+#   Array of additional command line arguments to pass to 'certbot renew'.
+# [*cron_ensure*]
+#   Intended state of the cron resource running certbot renew. Accepts 'present'
+#   or 'absent'. Default: 'absent'
+# [*cron_hour*]
+#   Optional string, integer or array of hour(s) the renewal command should run.
+#   E.g. '[0,12]' to execute at midnight and midday. Default: fqdn-seeded random
+#   hour.
+# [*cron_minute*]
+#   Optional string, integer or array of minute(s) the renewal command should
+#   run. E.g. 0 or '00' or [0,30]. Default: fqdn-seeded random minute.
+# [*cron_monthday*]
+#   Optional array of monthday(s) the renewal command should run. E.g. ['2-30/2']
+#   to run on even days. Default: Every day.
+#
+class letsencrypt::renew (
+  Variant[String, Array[String[1]]]         $pre_hook_commands    = $letsencrypt::renew_pre_hook_commands,
+  Variant[String, Array[String[1]]]         $post_hook_commands   = $letsencrypt::renew_post_hook_commands,
+  Variant[String, Array[String[1]]]         $deploy_hook_commands = $letsencrypt::renew_deploy_hook_commands,
+  Array[String[1]]                          $additional_args      = $letsencrypt::renew_additional_args,
+  Enum['present', 'absent']                 $cron_ensure          = $letsencrypt::renew_cron_ensure,
+  Variant[Integer[0,23], String, Array]     $cron_hour            = $letsencrypt::renew_cron_hour,
```

Comment:
can you also update those to enforce the datatypes within the array and the minimal string length?

Reply:
Turns out they were not handling arrays of integers either. Hopefully this is good style. Someone should make types for standard cron parameters in stdlib ...

```
  Variant[String[1], Array[String[1]]] $deploy_hook_commands = $letsencrypt::renew_deploy_hook_commands,
  ...
  Variant[Integer[0,23], String[1], Array[Variant[Integer[0,23], String[1]]]]
                                       $cron_hour            = $letsencrypt::renew_cron_hour,
  Variant[Integer[0,59], String[1], Array[Variant[Integer[0,59], String[1]]]]
                                       $cron_minute          = $letsencrypt::renew_cron_minute,
  Variant[Integer[0,31], String[1], Array[Variant[Integer[0,31], String[1]]]]
                                       $cron_monthday        = $letsencrypt::renew_cron_monthday,
```